proto/structured.proto

```diff
@@ -0,0 +1,67 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+syntax = "proto2";
+package cockroach.proto;
+option go_package = "proto";
+
+import "gogoproto/gogo.proto";
+import "cockroach/proto/api.proto";
+
+option (gogoproto.sizer_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+message TableDescriptorInner {
+  message ColumnDescriptor {
+    optional uint64 id = 1 [(gogoproto.nullable) = false];
+    optional string name = 2 [(gogoproto.nullable) = false];
+  }
+  message IndexDescriptor {
+    optional uint64 id = 1 [(gogoproto.nullable) = false];
+    optional string name = 2 [(gogoproto.nullable) = false];
+    // Each column_id refers to a column in the TableDescriptor's columns.
+    // Special care is taken to update this when deleting columns.
+    repeated uint64 column_ids = 3 [(gogoproto.nullable) = false];
+  }
+  repeated ColumnDescriptor columns = 1 [(gogoproto.nullable) = false];
+  repeated IndexDescriptor indexes = 2 [(gogoproto.nullable) = false];
+}
```

Comment:
How about defining ColumnDescriptor and IndexDescriptor and getting rid of TableDescriptorInner. It would mean adding the two repeated lists (columns and indexes) to the TableDescriptor below.